extensions/exposition/documentation/vary.md

```diff
@@ -0,0 +1,39 @@
+# HTTP request details
+
+The vary directives family provides the capability to include HTTP request details as input for an
+operation call.
+
+## TL;DR
+
+```yaml
+exposition:
+  /:
+    vary:languages: [en, fr]
+    GET:
+      vary:embed:
+        language: lang          # predefined embeddings
+        :x-access-token: token  # raw http header value
+      endpoint: texts.get
+```
+
+## Embeddings
+
+Request parts are embedded into the operation call according to the mapping
+defined by the `vary:embed` directive.
+The keys in the embedding map are the names of the embedding functions, and the values are the
+property names of the operation call input object.
+
+> If a property is already present in the input, the embedded value will overwrite its current
+> value.
+
+### Language
+
+The `language` embedding substitutes the most matching language code based on the `accept-language`
+request header and a list of supported languages defined by the `vary:languages` directive, and also
+adds `accept-language` to the `vary` HTTP response header value.
+
+### Raw headers
+
+Keys in the embedding map starting with a semicolon (:) are the names of HTTP request headers whose
+values are to be embedded into an operation call.
+The names of these headers are then included in the `vary` HTTP response header.
```

Comment:
As Toa is CORS enabled by default do we want to also add those headers to [Access-Control-Allow-Headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers) pre-flight response (or have a configuration enabling that)?

Reply:
Access-Control manages access to the **response** headers, while `vary` embeds values of the **request** headers.

I cannot see any reason to expect these headers in the **response**. It's technically impossible now (Toa doesn't provide the way to do that).

Reply:
To allow browser client to supply those headers `toa` must return them in `Access-Control-Allow-Headers` pre-flight response (`OPTIONS` request to the same endpoint where those headers will appear in the `Access-Control-Request-Headers`). If we don't add this then we should add a statement that this feature is not CORS compatible outside of headers allowed for browser requests by default.

Reply:
I didn't take my pills before the previous answer.

Yes, these headers must be included in the CORS allow list.

We'll have to find a convinient way to do that, since Directives are only running during the actual requests.

Reply:
We use Express and `cors` module currently and it [by default allows all headers](https://expressjs.com/en/resources/middleware/cors.html#configuration-options):

```md
allowedHeaders: Configures the Access-Control-Allow-Headers CORS header. Expects a comma-delimited string (ex: ‘Content-Type,Authorization’) or an array (ex: ['Content-Type', 'Authorization']). If not specified, defaults to reflecting the headers specified in the request’s
```

We just need to [enable pre-flight](https://expressjs.com/en/resources/middleware/cors.html#enabling-cors-pre-flight).

Reply:
Sounds great.

Reply:
@tinovyatkin I think it's only possible with disabled preflight request caching. AFAIR, [Access-Control-Max-Age](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age) is applied to the `origin`, not the URL.